visualize_classification_groups.py

```diff
@@ -0,0 +1,206 @@
+	# image_stream = BytesIO()
+	# Image.fromarray(res).save(image_stream,format="PNG")
+	slide.shapes.add_picture(image_stream, left, top ,height,width)
+	txBox = slide.shapes.add_textbox(left, Inches(1), width, height)
+	tf = txBox.text_frame
+	tf.text = comment
+
+def cartesian_product(*arrays):
+	la = len(arrays)
+	dtype = np.result_type(*arrays)
+	arr = np.empty([len(a) for a in arrays] + [la], dtype=dtype)
+	for i, a in enumerate(np.ix_(*arrays)):
+		arr[...,i] = a
+	return arr.reshape(-1, la)
+
+def generate_powerpoint(h5_path, model, pptx_save_path, img_transform, device, criteria_dict:dict):
+	dset = Dataset(h5_path, img_transform)
+	dloader = DataLoader(dset, batch_size=1, 
+								shuffle=True, num_workers=8,pin_memory=True)
+
+	# If model is set to None, we assume that predictions have already been generated and saved in the "prediction"
+	# dataset of the h5 file.
+	filters=tables.Filters(complevel=6, complib='zlib')
+	img_dtype = tables.UInt8Atom()
+	if (model != None):
+		predictions = []
+		for img, _, _ in tqdm(dloader):
+			img = img.to(device)
+			pred = model(img)
+			p=pred.detach().cpu().numpy()
+			predflat=np.argmax(p,axis=1).flatten()
+			predictions.append(predflat)
+			
+
+		# TODO something wrong here
+		with tables.open_file(h5_path, 'a') as f:
+			try:
+				predictions_dataset = f.create_carray(f.root, "predictions", img_dtype, np.array(predictions).shape)
+			except:
+				predictions_dataset = f.root.predictions
+
+			predictions_dataset[:] = predictions
+			
+	
+	# select indices for each category TN, FN, FP, TP
+	with tables.open_file(h5_path, 'r') as f:
+		gts = np.array(f.root.labels[:]).flatten()
+		preds = np.array(f.root.predictions[:]).flatten()
+		print(gts.shape)
+
+	ppt = Presentation()
+	cartesian_coords = cartesian_product(np.arange(5), np.arange(8))	# generate coordinates for the image grid
+
+	slidenames = criteria_dict.keys()
+	for slidename in slidenames:
+		criterion = criteria_dict[slidename]
+		inds = np.argwhere(np.logical_and(gts==criterion[0], preds==criterion[1])).flatten()	# find indices where the criterion is true
+		num_imgs = cartesian_coords.shape[0] if len(inds) >= cartesian_coords.shape[0] else len(inds) 
+		selected_inds = np.random.choice(inds, num_imgs, replace=False)				# choose a subset of max 30 elements from this set
+		print(f'Number of images in {slidename}: {len(inds)}')
+		blank_slide_layout = ppt.slide_layouts[6]
+		slide = ppt.slides.add_slide(blank_slide_layout)
+		txBox = slide.shapes.add_textbox(Inches(ppt.slide_height.inches/2), Inches(0.2), Inches(2), Inches(1))	# add slide title
+		txBox.text_frame.text = slidename
+		for j, ind in tqdm(enumerate(selected_inds)):
+			coord = cartesian_coords[j]
+			imgnew, label, imgold = dset[ind]
+			with tables.open_file(dset.fname, 'r') as f:	# we need to get the image_filename manually.
```

Comment:
here you're opening the table, but the table is already open in the background because of the dataloader. while for reading this shouldn't cause any problems (Though i have seen it with hdf5) this is considered bad practice as there may be some race conditions if the underlying library isn't threadsafe